src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/notifications/NotificationSender.scala

```diff
@@ -24,41 +24,73 @@ import za.co.absa.hyperdrive.trigger.models.{DagInstance, JobInstance, Notificat
 
 trait NotificationSender {
-  def sendNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit]
+  def createNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit]
+  def sendNotifications(): Unit
 }
 
 @Component
 class NotificationSenderImpl(notificationRuleService: NotificationRuleService, emailService: EmailService,
                              sparkConfig: SparkConfig, notificationConfig: NotificationConfig,
                              generalConfig: GeneralConfig) extends NotificationSender {
+  private case class Message(recipients: Seq[String], subject: String, text: String, attempts: Int)
+
   private val logger = LoggerFactory.getLogger(this.getClass)
   private val sender = notificationConfig.senderAddress
   private val notificationEnabled = notificationConfig.enabled
   private val environment = generalConfig.environment
   private val yarnBaseUrl = sparkConfig.hadoopResourceManagerUrlBase
   private val dateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME
+  private val messageQueue = new ConcurrentLinkedQueue[Message]
 
-  def sendNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit] = {
+  def createNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit] = {
     if (notificationEnabled) {
       notificationRuleService.getMatchingNotificationRules(dagInstance.workflowId, dagInstance.status).map {
-        case Some((rules, workflow)) => rules.foreach(rule => createMessageAndSend(rule, workflow, dagInstance, jobInstances))
-        case None => {
-                logger.debug(s"No rules matching workflow ID ${dagInstance.workflowId} with status ${dagInstance.status} found")
-                Future.successful()
-        }
+        case Some((rules, workflow)) => rules
+          .map(rule => createMessage(rule, workflow, dagInstance, jobInstances))
+          .foreach { message =>
+            logger.debug(s"Adding to queue message ${message.subject} from ${sender} to ${message.recipients}")
+            messageQueue.add(message)
+          }
+        case None =>
+          logger.debug(s"No rules matching workflow ID ${dagInstance.workflowId} with status ${dagInstance.status} found")
+          Future.successful()
       }
     } else {
       logger.debug(s"Attempting to send notifications for ${dagInstance}, but it is disabled")
       Future{}
     }
   }
 
-  private def createMessageAndSend(notificationRule: NotificationRule, workflow: Workflow, dagInstance: DagInstance,
-                                   jobInstances: Seq[JobInstance]): Unit = {
+  def sendNotifications(): Unit = {
+    import scala.collection.JavaConverters._
+    val messages = messageQueue.iterator().asScala.toList
+    messageQueue.removeAll(messages.asJava)
```

Comment:
I would rather not removeAll messages from messageQueue. You can have while with messageQueue.poll != null.

If there is an exception it will forget all removed messages.

Reply:
Yeah that's not good. Originally I didn't want to use the while loop to make sure that retried messages would only be retried in the next iteration of the scheduler to guarantee some delay. However, using only a single thread to send emails, it seems that there is a sufficient natural delay so it wouldn't be an issue.